pkg/providers/vsphere/virtualmachine/snapshot.go

```diff
@@ -0,0 +1,89 @@
+// © Broadcom. All Rights Reserved.
+// The term “Broadcom” refers to Broadcom Inc. and/or its subsidiaries.
+// SPDX-License-Identifier: Apache-2.0
+
+package virtualmachine
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
+
+	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha4"
+	vmopv1common "github.com/vmware-tanzu/vm-operator/api/v1alpha4/common"
+	pkgctx "github.com/vmware-tanzu/vm-operator/pkg/context"
+)
+
+// SnapshotArgs contains the options for createSnapshot.
+type SnapshotArgs struct {
+	VMCtx      pkgctx.VirtualMachineContext
+	VcVM       *object.VirtualMachine
+	VMSnapshot vmopv1.VirtualMachineSnapshot
+}
+
+func SnapshotVirtualMachine(args SnapshotArgs) (*types.ManagedObjectReference, error) {
+	obj := args.VMSnapshot
+	vm := args.VcVM
+	// find snapshot by name
+	snapMoRef, _ := vm.FindSnapshot(args.VMCtx, obj.Name)
+	if snapMoRef != nil {
+		// TODO: Handle revert to snapshot
+		// update vm.status with currentSnapshot
+		updateVMStatusCurrentSnapshot(args.VMCtx, obj)
+		// return early, snapshot found
+		return snapMoRef, nil
+	}
+
+	// if no snapshot was found, create it
+	args.VMCtx.Logger.Info("Creating Snapshot of VirtualMachine", "snapshot name", obj.Name)
+	snapMoRef, err := CreateSnapshot(args)
+	if err != nil {
+		args.VMCtx.Logger.Error(err, "failed to create snapshot for VM", "snapshot", obj.Name)
+		return nil, err
+	}
+
+	// update vm.status with currentSnapshot
+	updateVMStatusCurrentSnapshot(args.VMCtx, obj)
+	return snapMoRef, nil
+}
+
+func CreateSnapshot(args SnapshotArgs) (*types.ManagedObjectReference, error) {
+	snapObj := args.VMSnapshot
+	var quiesceSpec *types.VirtualMachineGuestQuiesceSpec
+	if quiesce := snapObj.Spec.Quiesce; quiesce != nil {
+		quiesceSpec = &types.VirtualMachineGuestQuiesceSpec{
+			Timeout: int32(quiesce.Timeout.Round(time.Minute).Minutes()),
+		}
+	}
+
+	t, err := args.VcVM.CreateSnapshotEx(args.VMCtx, snapObj.Name, snapObj.Spec.Description, snapObj.Spec.Memory, quiesceSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	// wait for task to finish
+	taskInfo, err := t.WaitForResult(args.VMCtx)
+	if err != nil {
+		if taskInfo != nil {
+			args.VMCtx.Logger.V(5).Error(err, "create snapshot task failed", "taskInfo", taskInfo)
+		}
```

Comment:
Not sure when the taskInfo would be nil and `err` will be not nil, but to handle that maybe log this line unconditionally?